numpy/core/src/multiarray/item_selection.c

```diff
@@ -1749,6 +1750,281 @@ PyArray_LexSort(PyObject *sort_keys, int axis)
     return NULL;
 }
 
+/*
+* Helper function for PyArray_KeySort
+* Sorts 1D slice based on argsort indices
+* takes in raw bytes of data, indices array
+* Number of elements, and stride of data
+* element size and temporary storage for an element.
+*/
+void reorder_by_index(char *data, char *current, npy_intp *index, npy_intp N, npy_intp stride, npy_intp elsize) {
+    // Reorder the indices and relevant values cyclically
+    for(npy_intp i = 0; i < N; i++){
+        /*Backup current element, end of loop place it in correct location.
+        *Stride here handles both aligned and non-aligned case
+        *When stride is same as elsize we are operating on aligned data.
+        */
+        memmove(current, data + (i*stride), elsize);
+        npy_intp current_idx = i;
```

Comment:
Should we do a quick check whether the array is already in place here? It probably doesn't matter much, but can skip the `memmove` above.  But, it also may make it more obvious that the approach is `O(N)`